### PR TITLE
Add Google Test and Mock post

### PR DESCRIPTION
**Pull request by fgs-pwright**
This post documents work I and a few other people did with Google Test and Google Mock earlier this year in support of adding them to the ModularCAP repo to test functionality in its `Launcher`.  Its intention is to document those experiences and provide some best practice recommendations for working with the framework.

### CODE REVIEW

File: _posts/2017-10-10-Google-Test-And-Google-Mock.md [hunk 1]
```diff
@@ -0,0 +1,157 @@
+---
+layout: post
+title: Google Test and Google Mock
+author: Phil Wright
+excerpt: Google's Test and Mock frameworks for C++ are powerful tools to support clean, object-oriented C++ applications
+
+---
+
+[Google Test](https://github.com/google/googletest#google-test) is a popular C++ unit testing framework developed by Google that can be used together with the closely related extension framework [Google Mock](https://github.com/google/googletest/blob/master/googlemock/README.md#google-mock) to test code that conforms to the SOLID principles for object-oriented software design. In order to follow best practices for testing while supporting a legacy application written in C++ and a new project in its early stages, FoxGuard's Security Engineering Team recently invested some time into incorporating these frameworks into the team's test pipeline alongside similar tools already used for .NET development. This post documents lessons learned during the setup of Google Test and Google Mock for the legacy application. Work done on the legacy application was in a Windows environment using Visual Studio for development and MSVC to compile the test and mock libraries and the application.
```
Comment (FGS-BMarkham) [inline _posts/2017-10-10-Google-Test-And-Google-Mock.md]: > … FoxGuard's Security Engineering Team …

While we don't have a clear canonical name for our team, "Software" probably belongs in there somewhere. Or you could say something like "our team". (I prefer the latter but if vagueness isn't your style please consider the former.)

File: _posts/2017-10-10-Google-Test-And-Google-Mock.md [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+
+## Setting Up the Libraries ##
+
+The recommended way to incorporate Google Test and Mock into a project is to compile the source code from [the googletest repository](https://github.com/google/googletest) and to link it into the project under test. The source files needed to build these two projects are in `googletest/src` and `googlemock/src`. Google provided Visual C++ project and solution files for building in `googletest/msvc` and `googlemock/msvc`, but the Google Test files were targeted at older versions of Visual Studio, so we made our own versions of these. We included `googletest` as a submodule within a `GoogleTest` folder in the repo housing the application, and we stored our custom solution files in this folder.
```
Comment (FGS-BMarkham) [inline _posts/2017-10-10-Google-Test-And-Google-Mock.md]: At this point, someone like me might be having trouble visualizing the file/folder structure. A diagram or screenshot of some sort might be helpful.

File: _posts/2017-10-10-Google-Test-And-Google-Mock.md [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+
+## Test Project Configuration ##
+
+A test project should be an executable project, and it will need additional include directories `$(SolutionDir)GoogleTest\googletest\googletest\include;$(SolutionDir)GoogleTest\googletest\googlemock\include;`. The Runtime Library for the test project and GTest and GMock needs to match the Runtime Library of the system under test.
```
Comment (FGS-BMarkham) [inline _posts/2017-10-10-Google-Test-And-Google-Mock.md]: I'm not sure that your capitalization of "Runtime Library" here is appropriate.

Comment (FGS-BMarkham) [inline _posts/2017-10-10-Google-Test-And-Google-Mock.md]: This is the first time you've shortened "Google Test" and "Google Mock" to "GTest" and "GMock", respectively, within this write-up. If these are common shortenings, consider introducing them when you originally mention the full names of the frameworks, in order to help the reader make the mental association. If you aren't going to consistently use these abbreviations during the remainder of the document, please consider not abbreviating them here.

File: _posts/2017-10-10-Google-Test-And-Google-Mock.md [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+(In Visual Studio: Show Configuration Properties > C/C++ > Code Generation)
```
Comment (FGS-BMarkham) [inline _posts/2017-10-10-Google-Test-And-Google-Mock.md]: A visual aide would probably be helpful here, such as a screenshot or GIF capture.

File: _posts/2017-10-10-Google-Test-And-Google-Mock.md [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+
+It is possible to create a test project whose SUT is an executable project, but this requires additional configuration as well as some additional attributes in SUT class declarations. Even with this additional configuration, we encountered bugs that were difficult to explain or address when running test code that wrote to managed memory, so we advise making the SUT a static library and referencing it from a separate executable project with minimal functionality, as needed.
+
+## Google Test ##
+
+Basic testing structure includes _tests_ and _assertions_. Tests do not have to be written as part of a test fixture, but it is good OO practice to do so.
+Google Test uses macros for tests and assertions. A test in simple, declarative code uses the `TEST()` macro, and a test as part of a fixture uses the `TEST_F()` macro. ([Example from the Google Test primer](https://github.com/google/googletest/blob/master/googletest/docs/Primer.md#test-fixtures-using-the-same-data-configuration-for-multiple-tests))
```
Comment (FGS-BMarkham) [inline _posts/2017-10-10-Google-Test-And-Google-Mock.md]: You've gone back to calling it "Google Test" after shortening it to "GTest" in the previous section. Consider using the shortened term consistently after its introduction.

Comment (FGS-BMarkham) [inline _posts/2017-10-10-Google-Test-And-Google-Mock.md]: Although hyperlinked and in a parenthetical, "Example from the Google Test primer" is an incomplete sentence fragment. Please consider making it a complete sentence (imperative mood).

File: _posts/2017-10-10-Google-Test-And-Google-Mock.md [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+A test fixture is written as a class that inherits `testing::Test`, and then each test that belongs to that fixture actually inherits the fixture class by way of the `TEST_F()` macro. (Show BitChecker.h / BitCheckerTests.cpp example in VS.)
```
Comment (FGS-BMarkham) [inline _posts/2017-10-10-Google-Test-And-Google-Mock.md]: The phrases "BitChecker.h" & "BitCheckerTests.cpp" are technical tokens in this concept, and should be demarcated with pairs of backticks (just like `TEST_F()`) for readability.

Comment (FGS-BMarkham) [inline _posts/2017-10-10-Google-Test-And-Google-Mock.md]: You've abbreviated "Visual Studio" here. I've found that this shortening, to "VS", doesn't resonate well with people who aren't used to using Visual Studio on a regular basis. Please consider not shortening it thusly.

File: _posts/2017-10-10-Google-Test-And-Google-Mock.md [hunk 7]
```diff
@@ -0,0 +1,157 @@
...
+        WaxIsRemaining_GivenWaxApplied_ReturnsTrue)
+    {
+        _waxApplicator.ApplyWax();
+
+        ASSERT_TRUE(_waxApplicator.WaxIsRemaining());
+    }
+}
+```
+
+Google Test actual mangles the test names using underscores internally (something like "`WaxApplicatorTests_WaxIsRemaining_Given...`") so there are general uniqueness concerns, but we should not need to worry about them, as long as we do not include any underscores in our test fixture names. That said, the documentation recommends against underscores altogether in the interest future-proofing your tests, as it makes no guarantees that a future change might not change how names are managed internally.
```
Comment (FGS-BMarkham) [inline _posts/2017-10-10-Google-Test-And-Google-Mock.md]: This sentence is a bit long and contains a few disjunct thoughts. Consider breaking it up please.

File: _posts/2017-10-10-Google-Test-And-Google-Mock.md [hunk 8]
```diff
@@ -0,0 +1,157 @@
...
+    public:
+        virtual ~ICarFacade() {}
+        virtual WaxType GetPreferredWax(int) const = 0;
+    };
+}
+```
+
+The reason the destructor is defined explicitly is that if a mock were to be derived from a class without a virtual destructor, then the derived destructor would not be called when a mock object is deleted through a pointer to the base class.
+
+Mock class declaration uses some strange syntax. All mock methods should be public. The docs give clear instructions for how to write these, and it requires use of old function pointer syntax.
```
Comment (FGS-BMarkham) [inline _posts/2017-10-10-Google-Test-And-Google-Mock.md]: The scoping "public" is a technical token and should probably be demarcated with markdown backticks.

Comment (FGS-BMarkham) [inline _posts/2017-10-10-Google-Test-And-Google-Mock.md]: What make the function pointer syntax "old"? Is this a judgment call or descriptor that would be more familiar to a regular C/C++ developer?

File: _posts/2017-10-10-Google-Test-And-Google-Mock.md [hunk 9]
```diff
@@ -0,0 +1,157 @@
...
+        ASSERT_TRUE(_waxApplicator.WaxIsRemaining());
+    }
+}
+```
+
+Google Test actual mangles the test names using underscores internally (something like "`WaxApplicatorTests_WaxIsRemaining_Given...`") so there are general uniqueness concerns, but we should not need to worry about them, as long as we do not include any underscores in our test fixture names. That said, the documentation recommends against underscores altogether in the interest future-proofing your tests, as it makes no guarantees that a future change might not change how names are managed internally.
+
+## Google Mock ##
+
+A mock should inherit from an abstract class, generally, that defines a virtual destructor explicitly and defines pure virtual functions for any methods that will be needed by the mock objects. (This is not a strict requirement, but it avoids some complications.) We followed the [Google C++ Style Guide](https://google.github.io/styleguide/cppguide.html) specifications for creating _interfaces_ (a construct, not a language feature) to define these base classes.
```
Comment (FGS-BMarkham) [inline _posts/2017-10-10-Google-Test-And-Google-Mock.md]: The modifier "virtual" is a technical token and should probably be demarcated with markdown backticks.

File: _posts/2017-10-10-Google-Test-And-Google-Mock.md [hunk 10]
```diff
@@ -0,0 +1,157 @@
...
+}
+```
+
+## Automating Test Run for Pull Requests Using Jenkins ##
+
+Incorporating Google Test into the continuous integration pipeline the team already had in place was fairly straightforward, since the framework is based on xUnit. The [Jenkins xUnit plugin](https://plugins.jenkins.io/xunit)'s `archiveXUnit` tool includes a utility for reading Google Test reports. Building our test project produced an executable, and producing the necessary test report required running the executable and passing it the argument `--gtest_output="xml:{outputFileName}"`. This proved a little more straightforward than creating reports for our .NET-based NUnit test projects, each of which produces a library that is passed as an argument to the NUnit console runner we have installed. The C++ test project requires that the Google Test resources be present when the test project is built, whereas the .NET projects utilizes the NUnit test resources when the test script is run. Since different Jenkins plugins were used for reading these test reports (there is an NUnit-specific plugin), we patterned the report filenames differently so as to be able to specify a filename pattern for each plugin to target.
```
Comment (FGS-BMarkham) [inline _posts/2017-10-10-Google-Test-And-Google-Mock.md]: I wouldn't go into implementation details of our .NET & NUnit pipeline. It's a bit off topic, and without further explanation/description it might actually be more of a distraction to the reader than a helpful comparison. Also, I'm confused what you mean by patterned filenames. Visual comparisons might be helpful - but like I said, it might also be a distraction.

File: _posts/2017-10-10-Google-Test-And-Google-Mock.md [hunk 11]
```diff
@@ -0,0 +1,157 @@
...
+
+Building a solution that includes Google Test on our Jenkins server required us to include the `submoduleOptions{recursive(true)}` option in our SCM call, so that Google Test would be cloned recursively inside the solution that required it for reference when building the test project. Outside this change, though, no additional modifications were needed to our Jenkins configuration to incorporate Google Test.
```
Comment (FGS-BMarkham) [inline _posts/2017-10-10-Google-Test-And-Google-Mock.md]: Even Jenkins fans aren't going to know what `submoduleOptions{recursive(true)}` means. Jenkins Job DSL is not particularly prevalent (especially since Jenkins has recently introduced an official competing alternative). Consider showing what the job configuration UI looks like for this kind of configuration change.

File: _posts/2017-10-10-Google-Test-And-Google-Mock.md (hunk 7 above)
Comment (FGS-BMarkham) [reply _posts/2017-10-10-Google-Test-And-Google-Mock.md]: "Google Test actual mangles" -> "Google Test _actually_ mangles"

**Comment by FGS-BMarkham** [comment]
There are a handful of terms bandied about that probably need to be defined or hyperlinked on first use. YMMV, but here is the list of what caught my eye:
- unit testing
- SOLID principles
- SUT
- mock

File: _posts/2017-10-10-Google-Test-And-Google-Mock.md [hunk 12]
```diff
@@ -0,0 +1,157 @@
...
+
+Call verification is done by setting an expectation, which is then verified when the mock object goes into garbage collection (recall that Google Mock handles destroying the mock -- this is part of it). Note that an expectation will only be satisfied if the method call occurs *after* the expectation is set. For example, if you instantiate a class under test with a mock object dependency in the test fixture's constructor, and that class's constructor makes a call to the mock object, an `EXPECT_CALL()` for that method call will go unsatisfied. You can do setup of a mock using the `ON_CALL()` macro. Different matchers can be used in `EXPECT_CALL()`, eg., `testing::_`, `testing::Pointee()`, `testing::StrEq()`.
```
Comment (fgs-dbudwin) [inline _posts/2017-10-10-Google-Test-And-Google-Mock.md]: Let's put the em-dash in here:

" (recall that Google Mock handles destroying the mock—this is part of it)"

File: _posts/2017-10-10-Google-Test-And-Google-Mock.md (hunk 11 above)
Comment (fgs-dbudwin) [inline _posts/2017-10-10-Google-Test-And-Google-Mock.md]: "though" feels like a filler word.

"Outside this change, no additional modifications..."

File: _posts/2017-10-10-Google-Test-And-Google-Mock.md [hunk 13]
```diff
@@ -0,0 +1,157 @@
...
+
+Working with a `shared_ptr` to a mock object can cause some complications, because Google Test wants to take care of garbage collection on its own. This can cause memory access errors that would not occur when working with raw pointers to these mocks. Consequently, if the class under test has a dependency on a `shared_ptr` to an interface that will be mocked, the shared pointer should be created referencing the mock with a trivial destructor.
+
+```cpp
+#include "gtest/gtest.h"
```
Comment (fgs-dbudwin) [inline _posts/2017-10-10-Google-Test-And-Google-Mock.md]:
```
#include "gtest/gtest.h"

#include "MockCarFacade.h"
#include "WaxApplicator.h"
```

File: _posts/2017-10-10-Google-Test-And-Google-Mock.md [hunk 14]
```diff
@@ -0,0 +1,157 @@
...
+    };
+}
+```
+
+The reason the destructor is defined explicitly is that if a mock were to be derived from a class without a virtual destructor, then the derived destructor would not be called when a mock object is deleted through a pointer to the base class.
+
+Mock class declaration uses some strange syntax. All mock methods should be public. The docs give clear instructions for how to write these, and it requires use of old function pointer syntax.
+
+```cpp
+#include "gtest/gtest.h"
```
Comment (fgs-dbudwin) [inline _posts/2017-10-10-Google-Test-And-Google-Mock.md]:
```
#include "gmock/gmock.h"
#include "gtest/gtest.h"

#include "ICarFacade.h"
```

File: _posts/2017-10-10-Google-Test-And-Google-Mock.md [hunk 15]
```diff
@@ -0,0 +1,157 @@
...
+        explicit WaxApplicator();
+        void ApplyWax() const override;
+        bool WaxIsRemaining() const override;
+        void BuffSurface() const override;
+    }
+}
+```
+
+```cpp
+#include "gtest/gtest.h"
```
Comment (fgs-dbudwin) [inline _posts/2017-10-10-Google-Test-And-Google-Mock.md]:
```
#include "gtest/gtest.h"

#include "WaxApplicator.h"
```

File: _posts/2017-10-10-Google-Test-And-Google-Mock.md [hunk 16]
```diff
@@ -0,0 +1,157 @@
...
+
+Call verification is done by setting an expectation, which is then verified when the mock object goes into garbage collection (recall that Google Mock handles destroying the mock -- this is part of it). Note that an expectation will only be satisfied if the method call occurs *after* the expectation is set. For example, if you instantiate a class under test with a mock object dependency in the test fixture's constructor, and that class's constructor makes a call to the mock object, an `EXPECT_CALL()` for that method call will go unsatisfied. You can do setup of a mock using the `ON_CALL()` macro. Different matchers can be used in `EXPECT_CALL()`, eg., `testing::_`, `testing::Pointee()`, `testing::StrEq()`.
+
+Working with a `shared_ptr` to a mock object can cause some complications, because Google Test wants to take care of garbage collection on its own. This can cause memory access errors that would not occur when working with raw pointers to these mocks. Consequently, if the class under test has a dependency on a `shared_ptr` to an interface that will be mocked, the shared pointer should be created referencing the mock with a trivial destructor.
```
Comment (fgsahoward) [inline _posts/2017-10-10-Google-Test-And-Google-Mock.md]: I'm not 100% sold that this description is accurate of the code you show below. In the example, Google Mock is only creating an instance of the test class, which in turn allocates memory (on the stack) for the `MockCarFacade _mockCarFacade` instance variable. Then the constructor creates a `shared_ptr<ICarFacade>` which points to the `_mockCarFacade`. This means that when the destructor of `WaxApplicatorTests` is called, it will call the destructor of each of its instance variables, namely of  `WaxApplicator _waxApplicator`. This in turn turtles all the way down until the destructor of the `shared_ptr` is called, which by default will call `delete _internalPointer`. That `delete` call is invalid since the pointer it holds points to stack memory, and not heap memory allocated with `new`. 

That is a long winded way of saying that the default `deleter` parameter of `std::shared_ptr` is `[] (auto p) { delete p; }`, which is invalid in this use case, because `p`  is on the stack and not heap allocated using `new`. In short, it isn't Google Test doing memory management that causes this problem, it's the way in which you have to create Mocks with Google Test that causes us to have to dance this dance.

File: _posts/2017-10-10-Google-Test-And-Google-Mock.md (hunk 8 above)
Comment (fgsahoward) [reply _posts/2017-10-10-Google-Test-And-Google-Mock.md]: Technically I don't think these are function pointers at all, they read more like function signatures. An old style function pointer would look more like `size_t (*count) (int*)`

File: _posts/2017-10-10-Google-Test-And-Google-Mock.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (fgs-pwright) [reply _posts/2017-10-10-Google-Test-And-Google-Mock.md]: > Google provided _a_ Visual C++ project and solution

No, there were multiple project files and multiple solution files.

File: _posts/2017-10-10-Google-Test-And-Google-Mock.md (hunk 3 above)
Comment (fgs-pwright) [reply _posts/2017-10-10-Google-Test-And-Google-Mock.md]: I had them abbreviated in my shorthand notes and copy-pasted from those.  I thought I took out all occurrences but obviously missed some!  Thanks.

File: _posts/2017-10-10-Google-Test-And-Google-Mock.md (hunk 9 above)
[the remark this answers is not in the document]
Comment (fgs-pwright) [reply _posts/2017-10-10-Google-Test-And-Google-Mock.md]: I'm fine removing these parentheses.

Comment (fgs-pwright) [reply _posts/2017-10-10-Google-Test-And-Google-Mock.md]: I'd like to leave this sentence as it is, for a couple reasons.  First, in the [Google Style Guide section on interfaces](https://google.github.io/styleguide/cppguide.html#Interfaces), the term is not highlighted in this context.  (Though, I acknowledge that the guide is not completely consistent in its use of syntax highlighting.)  Second, though, I use the term "pure virtual," which is not a technical token, but is denoted by the inclusion of the `virtual` keyword and `= 0` at the end of the method declaration.  I would find it awkward to highlight the first use of "virtual" and not highlight the same word in the term "pure virtual."

Comment (FGS-BMarkham) [reply _posts/2017-10-10-Google-Test-And-Google-Mock.md]: Alright, fair enough.

File: _posts/2017-10-10-Google-Test-And-Google-Mock.md (hunk 8 above)
Comment (fgs-pwright) [reply _posts/2017-10-10-Google-Test-And-Google-Mock.md]: Oof, I remember reading somewhere that some of the syntax used to define mock methods had some similarity to function pointer syntax, and I can't seem to track that down now.  I'll take this out.

File: _posts/2017-10-10-Google-Test-And-Google-Mock.md (hunk 16 above)
Comment (fgs-pwright) [reply _posts/2017-10-10-Google-Test-And-Google-Mock.md]: Thanks.  I'll reword this to be more accurate.  (Or maybe just to be accurate, period.)

File: _posts/2017-10-10-Google-Test-And-Google-Mock.md (hunk 11 above)
Comment (fgs-pwright) [reply _posts/2017-10-10-Google-Test-And-Google-Mock.md]: Yeah, I don't think I ended this well.  I just realized I was done and didn't have a conclusion.

This reminds me of my compositions from elementary school: "Here is what I did with my summer vacation.  Over my summer vacation I went to the pool a lot.  It was fun.  That is what I did with my summer vacation."

Maybe I'll try to make it end less abruptly.

Comment (fgs-pwright) [reply _posts/2017-10-10-Google-Test-And-Google-Mock.md]: Oh, got it.

Comment (fgs-dbudwin) [reply _posts/2017-10-10-Google-Test-And-Google-Mock.md]: Awesome.  That would be great.  Looking forward to this change.  Let me know if I can help.  You're the best.

File: _posts/2017-10-10-Google-Test-And-Google-Mock.md (hunk 6 above)
Comment (fgs-pwright) [reply _posts/2017-10-10-Google-Test-And-Google-Mock.md]: Oops, that parenthetical note should have been removed!

**Comment by fgs-pwright** [comment]
Thanks, all, for the feedback you've given!  I pushed a `fixup` that I believe addresses all of it.

File: _posts/2017-10-10-Google-Test-And-Google-Mock.md [hunk 17]
```diff
@@ -0,0 +1,173 @@
...
+
+## Google Test ##
+
+Basic testing structure includes _tests_ and _assertions_. Tests do not have to be written as part of a test fixture, but it is good OO practice to do so.
+Google Test uses macros for tests and assertions. A test in simple, declarative code uses the `TEST()` macro, and a test as part of a fixture uses the `TEST_F()` macro. ([See example from the Google Test primer.](https://github.com/google/googletest/blob/master/googletest/docs/Primer.md#test-fixtures-using-the-same-data-configuration-for-multiple-tests))
+A test fixture is written as a class that inherits `testing::Test`, and then each test that belongs to that fixture actually inherits the fixture class by way of the `TEST_F()` macro.
```
Comment (fgsahoward) [inline _posts/2017-10-10-Google-Test-And-Google-Mock.md]: <pedantic>
This sentence seems to imply that your test functions are inheriting a test class. I'm not sure what it means for a function to inherit a class. Can you clear this up a bit?
</pedantic>

Comment (fgs-dbudwin) [reply _posts/2017-10-10-Google-Test-And-Google-Mock.md]: https://media.giphy.com/media/WyOpdUnmsc9Da/giphy.gif

File: _posts/2017-10-10-Google-Test-And-Google-Mock.md [hunk 18]
```diff
@@ -0,0 +1,173 @@
...
+
+![Configure the runtime library to match across projects]({{ site.baseurl }}/images/2017-10-10-Google-Test-And-Google-Mock/Configure_Runtime_Library.gif)
+
+It is possible to create a test project whose SUT is an executable project, but this requires additional configuration as well as some additional attributes in SUT class declarations. Even with this additional configuration, we encountered bugs that were difficult to explain or address when running test code that wrote to managed memory, so we advise making the SUT a static library and referencing it from a separate executable project with minimal functionality, as needed.
+
+## Google Test ##
+
+Basic testing structure includes _tests_ and _assertions_. Tests do not have to be written as part of a test fixture, but it is good OO practice to do so.
```
Comment (fgs-dbudwin) [inline _posts/2017-10-10-Google-Test-And-Google-Mock.md]: Why are test and assertions italicized at first, but not later?

File: _posts/2017-10-10-Google-Test-And-Google-Mock.md [hunk 19]
```diff
@@ -0,0 +1,173 @@
...
+Google Test uses macros for tests and assertions. A test in simple, declarative code uses the `TEST()` macro, and a test as part of a fixture uses the `TEST_F()` macro. ([See example from the Google Test primer.](https://github.com/google/googletest/blob/master/googletest/docs/Primer.md#test-fixtures-using-the-same-data-configuration-for-multiple-tests))
```
Comment (fgs-dbudwin) [inline _posts/2017-10-10-Google-Test-And-Google-Mock.md]: I think the parenthetical statement feels awkward sitting by itself trying to be it's own sentence.  I would suggest either making it its own sentence or use the following:

Google Test uses macros for tests and assertions. A test in simple, declarative code uses the `TEST()` macro, and a test as part of a fixture uses the `TEST_F()` macro ([see example from the Google Test primer](https://github.com/google/googletest/blob/master/googletest/docs/Primer.md#test-fixtures-using-the-same-data-configuration-for-multiple-tests)).

File: _posts/2017-10-10-Google-Test-And-Google-Mock.md [hunk 20]
```diff
@@ -0,0 +1,173 @@
...
+        void BuffSurface() const override;
+    }
+}
+```
+
+```cpp
+#include "gtest/gtest.h"
+
+#include "WaxApplicator.h"
+// ...
```
Comment (fgs-dbudwin) [inline _posts/2017-10-10-Google-Test-And-Google-Mock.md]: I would remove this comment, I don't think it adds anything besides some clutter.

File: _posts/2017-10-10-Google-Test-And-Google-Mock.md [hunk 21]
```diff
@@ -0,0 +1,173 @@
...
+        WaxIsRemaining_GivenWaxApplied_ReturnsTrue)
+    {
+        _waxApplicator.ApplyWax();
+
+        ASSERT_TRUE(_waxApplicator.WaxIsRemaining());
+    }
+}
+```
+
+Internally, through macro expansion, Google Test will mangle test names using underscores (eg., something like "`WaxApplicatorTests_WaxIsRemaining_Given...`"). This could create name uniqueness problems, if both test fixtures and test names included underscores. Though the documentation is not quite clear on the matter, Google Test maintainers have stated that it is acceptable to use underscores in test names, but that underscores should be avoided in test fixture names. ([See discussion on this matter.](https://groups.google.com/forum/#!topic/googletestframework/N5A07bgEvp4))
```
Comment (fgs-dbudwin) [inline _posts/2017-10-10-Google-Test-And-Google-Mock.md]: I think if you use "eg." then the "something" feels wordy, also I don't think quotes around the class name are needed.  How about this:

Internally, through macro expansion, Google Test will mangle test names using underscores (eg., `WaxApplicatorTests_WaxIsRemaining_Given...`).

Comment (fgs-dbudwin) [reply _posts/2017-10-10-Google-Test-And-Google-Mock.md]: I would take the note at the end: "See discussion on this matter" and take that link and apply it to "Google Test maintainers have stated" and remove the parenthetical statement like the following:

Though the documentation is not quite clear on the matter, [Google Test maintainers have stated](https://groups.google.com/forum/#!topic/googletestframework/N5A07bgEvp4) that it is acceptable to use underscores in test names, but that underscores should be avoided in test fixture names.

File: _posts/2017-10-10-Google-Test-And-Google-Mock.md [hunk 22]
```diff
@@ -0,0 +1,173 @@
...
+
+## Google Mock ##
+
+A mock should inherit from an abstract class, generally, that defines a virtual destructor explicitly and defines pure virtual functions for any methods that will be needed by the mock objects. This is not a strict requirement, but it avoids some complications. We followed the [Google C++ Style Guide](https://google.github.io/styleguide/cppguide.html) specifications for creating _interfaces_ (a construct, not a language feature) to define these base classes.
```
Comment (fgs-dbudwin) [inline _posts/2017-10-10-Google-Test-And-Google-Mock.md]: "Generally" feels like a filler word to me.  I think it can be safely removed.  If that's not always the case (as "generally" implies), I would specifically mention the 1% of times you wouldn't do it the way you mentioned.

File: _posts/2017-10-10-Google-Test-And-Google-Mock.md (hunk 17 above)
Comment (fgs-pwright) [reply _posts/2017-10-10-Google-Test-And-Google-Mock.md]: :D I can see how this could come across that way.  I'll clarify this a little!

All of this come from notes I made for a presentation I did with the team way back months ago, and in the notes there were clear gaps that I intended to fill in verbally.  I tried to fill all those in, but this looks like one I missed.

File: _posts/2017-10-10-Google-Test-And-Google-Mock.md (hunk 22 above)
Comment (fgs-dbudwin) [reply _posts/2017-10-10-Google-Test-And-Google-Mock.md]: I think some things should be singular:

"A mock should inherit from an abstract class that defines a virtual destructor explicitly and defines a pure virtual function for any method that will be needed by the mock objects."

Comment (fgs-dbudwin) [reply _posts/2017-10-10-Google-Test-And-Google-Mock.md]: Why is "interfaces" italicized?

File: _posts/2017-10-10-Google-Test-And-Google-Mock.md (hunk 17 above)
Comment (fgs-pwright) [reply _posts/2017-10-10-Google-Test-And-Google-Mock.md]: > A test fixture is written as a class that inherits `testing::Test`, and then each test that belongs to that fixture is expanded by the `TEST_F()` macro into a class that inherits the fixture class.

How does this sound, @fgsahoward ?

File: _posts/2017-10-10-Google-Test-And-Google-Mock.md [hunk 23]
```diff
@@ -0,0 +1,173 @@
...
+        virtual WaxType GetPreferredWax(int) const = 0;
+    };
+}
+```
+
+The reason the destructor is defined explicitly is that if a mock were to be derived from a class without a virtual destructor, then the derived destructor would not be called when a mock object is deleted through a pointer to the base class.
+
+![Memory leaks are bad]({{ site.baseurl }}/images/2017-10-10-Google-Test-And-Google-Mock/Memory_Leaks.gif)
+
+Mock class declaration uses some strange syntax, but the docs give clear instructions for how to write mock method declarations [here](https://github.com/google/googletest/blob/master/googlemock/docs/ForDummies.md#how-to-define-it).
```
Comment (fgs-dbudwin) [inline _posts/2017-10-10-Google-Test-And-Google-Mock.md]: I don't like when things say "click here."  We should be more expressive with our wordage to make the link more subtle which is tricky sometimes.  How about (note I changed some words a bit):

Mock class declarations use some strange syntax, but the [documentation gives clear instructions for how to write mock method declarations](https://github.com/google/googletest/blob/master/googlemock/docs/ForDummies.md#how-to-define-it).

File: _posts/2017-10-10-Google-Test-And-Google-Mock.md [hunk 24]
```diff
@@ -0,0 +1,173 @@
...
+
+Building a solution that includes Google Test on our Jenkins server required us to configure our source code management with the "Recursively update submodules" option, so that Google Test would be cloned recursively inside the solution that required it for reference when building the test project. Outside this change, no additional modifications were needed to our Jenkins configuration to incorporate Google Test.
+
+![Jenkins SCM configuration]({{ site.baseurl }}/images/2017-10-10-Google-Test-And-Google-Mock/Jenkins_Job_Configuration_Recursive_Submodules.PNG)
+
+## Conclusion ##
+
+We found Google Test and Google Mock to be a fully-featured pair of frameworks that we could integrate into our development process and continuous integration pipeline without a great deal of effort. The frameworks integrate well with many of the tools we use on a regular basis, including Visual Studio, ReSharper's unit test runner, and Jenkins. The frameworks are well-documented with many examples and guides targeted at beginners and advanced users, and they seem to have a fairly substantial community of users. We are looking forward to using Google Test and Google Mock on future projects!
```
Comment (fgs-dbudwin) [inline _posts/2017-10-10-Google-Test-And-Google-Mock.md]: Can you link to an example of the begging guide and an advanced guide?

File: _posts/2017-10-10-Google-Test-And-Google-Mock.md [hunk 25]
```diff
@@ -0,0 +1,173 @@
...
+
+![The 'googletest' submodule and upgraded project files within our solution]({{ site.baseurl }}/images/2017-10-10-Google-Test-And-Google-Mock/Google_Test_Within_Our_Solution.PNG)
+
+## Test Project Configuration ##
+
+A test project should be an executable project, and it will need additional include directories `$(SolutionDir)GoogleTest\googletest\googletest\include;$(SolutionDir)GoogleTest\googletest\googlemock\include;`. The runtime library that is specified for the test project, the one specified for Google Test, and the one for Google Mock all need to match the runtime library specified for the system under test, or SUT.
```
[the remark this answers is not in the document]
Comment (fgs-pwright) [reply _posts/2017-10-10-Google-Test-And-Google-Mock.md]: Oh, I like that.  I'll make these changes.

File: _posts/2017-10-10-Google-Test-And-Google-Mock.md (hunk 18 above)
Comment (fgs-pwright) [reply _posts/2017-10-10-Google-Test-And-Google-Mock.md]: http://blog.apastyle.org/apastyle/2015/04/using-italics-for-technical-or-key-terms.html

File: _posts/2017-10-10-Google-Test-And-Google-Mock.md (hunk 22 above)
Comment (fgs-pwright) [reply _posts/2017-10-10-Google-Test-And-Google-Mock.md]: > Why is "interfaces" italicized?

http://blog.apastyle.org/apastyle/2015/04/using-italics-for-technical-or-key-terms.html was guiding this, too.

Comment (fgs-pwright) [reply _posts/2017-10-10-Google-Test-And-Google-Mock.md]: > "Generally" feels like a filler word to me. I think it can be safely removed.

I'd like to leave it.  The sentence following it qualifies it, and I think the second sentence would sound like it's contradicting the first one if I took "generally" out.

File: _posts/2017-10-10-Google-Test-And-Google-Mock.md [hunk 26]
```diff
@@ -0,0 +1,177 @@
...
+    };
+
+    TEST_F(WaxApplicatorTests,
+        WaxIsRemaining_GivenNoWaxApplied_ReturnsTrue)
+    {
+        ASSERT_FALSE(_waxApplicator.WaxIsRemaining());
+    }
+
+    TEST_F(WaxApplicatorTests,
+        WaxIsRemaining_GivenWaxApplied_ReturnsTrue)
```
Comment (FGS-JTKuhn) [inline _posts/2017-10-10-Google-Test-And-Google-Mock.md]: Your test methods are named wrongly here. I think the one above should be `_ReturnsFalse`.

Comment (fgs-pwright) [reply _posts/2017-10-10-Google-Test-And-Google-Mock.md]: Thanks, good catch!

I have corrected the test name in a `fixup`.

File: _posts/2017-10-10-Google-Test-And-Google-Mock.md [hunk 27]
```diff
@@ -0,0 +1,177 @@
...
+        virtual WaxType GetPreferredWax(int) const = 0;
+    };
+}
+```
+
+The reason the destructor is defined explicitly is that if a mock were to be derived from a class without a virtual destructor, then the derived destructor would not be called when a mock object is deleted through a pointer to the base class.
+
+![Memory leaks are bad]({{ site.baseurl }}/images/2017-10-10-Google-Test-And-Google-Mock/Memory_Leaks.gif)
+
+Mock class declaration uses some strange syntax, but [the documentation give clear instructions for how to write mock method declarations](https://github.com/google/googletest/blob/master/googlemock/docs/ForDummies.md#how-to-define-it).
```
Comment (fgs-dbudwin) [inline _posts/2017-10-10-Google-Test-And-Google-Mock.md]: I think this should be "the documentation gives clear instructions"

Comment (fgs-pwright) [reply _posts/2017-10-10-Google-Test-And-Google-Mock.md]: Good catch!  I'll correct this and :squash:.

**Comment by fgs-pwright** [comment]
Has been :squash:ed

**Comment by fgs-dbudwin** [comment]
image

**Comment by fgs-dbudwin** [comment]
@fgs-pwright Does this need to be rebased?

**Comment by fgs-pwright** [comment]
No, I rebased it when I squashed it.  Its parent commit is 098c41f.

**Comment by fgs-dbudwin** [comment]
It's asking me to "Squash and merge" which seems wrong.

image
